crates/biome_js_analyze/src/lint/nursery/use_at_index.rs

```diff
@@ -0,0 +1,717 @@
+}
+
+/// When using this expression in other operations, enclose it in parentheses as needed.
+fn overwrap_parentheses_expression(node: &AnyJsExpression) -> Option<AnyJsExpression> {
+    match node {
+        AnyJsExpression::JsArrayExpression(exp) => {
+            Some(AnyJsExpression::JsArrayExpression(exp.clone()))
+        }
+        AnyJsExpression::JsCallExpression(exp) => {
+            Some(AnyJsExpression::JsCallExpression(exp.clone()))
+        }
+        AnyJsExpression::JsComputedMemberExpression(exp) => {
+            Some(AnyJsExpression::JsComputedMemberExpression(exp.clone()))
+        }
+        AnyJsExpression::JsIdentifierExpression(exp) => Some(
+            AnyJsExpression::JsIdentifierExpression(exp.clone().trim_trivia()?),
+        ),
+        AnyJsExpression::JsParenthesizedExpression(exp) => {
+            Some(AnyJsExpression::JsParenthesizedExpression(exp.clone()))
+        }
+        AnyJsExpression::JsStaticMemberExpression(exp) => {
+            Some(AnyJsExpression::JsStaticMemberExpression(exp.clone()))
+        }
+        _ => Some(AnyJsExpression::JsParenthesizedExpression(
+            make::js_parenthesized_expression(
+                make::token(T!['(']),
+                node.clone(),
+                make::token(T![')']),
+            ),
+        )),
+    }
+}
+
+/// If the node is a length method, it returns the object of interest.
+fn get_length_node(node: &AnyJsExpression) -> Option<AnyJsExpression> {
+    let AnyJsExpression::JsStaticMemberExpression(node) = node else {
+        return None;
+    };
+    let member_name = node.member().ok()?;
+    let member_name = member_name
+        .as_js_name()?
+        .value_token()
+        .ok()?
+        .token_text_trimmed();
+    if member_name.text() != "length" {
+        return None;
+    }
+    node.object().ok()
+}
+
+/// AnyJsExpressiion -> Some(i64) if the expression is an integer literal, otherwise None.
+fn get_integer_from_literal(node: &AnyJsExpression) -> Option<i64> {
+    if let AnyJsExpression::JsUnaryExpression(unary) = node {
+        let token = unary.operator_token().ok()?;
+        if token.kind() != T![-] {
+            return None;
+        }
+        return get_integer_from_literal(&unary.argument().ok()?.omit_parentheses())
+            .map(|num| -num);
+    }
+    let AnyJsExpression::AnyJsLiteralExpression(AnyJsLiteralExpression::JsNumberLiteralExpression(
+        number,
+    )) = node
+    else {
+        return None;
+    };
+    let number = number.as_number()?;
+    if number.fract() == 0.0 {
+        Some(i64::try_from(number as i128).ok()?)
+    } else {
+        None
+    }
+}
+
+/// Retrieve the value subtracted from the subtraction expression.
+/// # Examples
+/// ```js
+///    a - b // => Some((a, [b]))
+///    a - b - c // => Some((a, [b, c]))
+/// ```
+fn split_minus_binary_expressions(
+    mut expression: AnyJsExpression,
+) -> Option<(AnyJsExpression, Vec<AnyJsExpression>)> {
+    let mut right_list = vec![];
```

Comment:
We could avoid allocating a `Vec` by using [std::iter::successors](https://doc.rust-lang.org/std/iter/fn.successors.html).
We could clone the iterator at call site if we need to iterate over it several times.

This is a refactor that we can delay to a future PR.